File: basics/node-programs/09_html_rendering.js
const http = require('http');

/*
	NODE.JS PROGRAM 9: HTML RENDERING & TEMPLATES
	Concepts:
	- Serving HTML strings from a node server.
	- Using template literals (ES6) for dynamic content in HTML string.
	- Loop rendering in a string-based template.
*/

// PageData (equivalent to the Go struct)
const pageData = {
	title: "Node.js Programming Language",
	author: "Ryan Dahl",
	language: "JavaScript (V8)",
	features: ["Non-blocking I/O", "V8 Engine", "NPM Ecosystem", "Unified Language (Front/Back)"]
};

const server = http.createServer((req, res) => {
	// Set correct MIME type to render text as HTML
	res.setHeader('Content-Type', 'text/html');
	
	// Create the HTML template string
	const html = `
	<!DOCTYPE html>
	<html>
	<head>
		<title>${pageData.title}</title>
		<style>
			body { font-family: Arial, sans-serif; padding: 2em; background-color: #f4f4f4; }
			h1 { color: #007d9c; }
			ul { color: #333; }
		</style>
	</head>
	<body>
		<h1>${pageData.title}</h1>
		<p><strong>Developed by:</strong> ${pageData.author}</p>
		<p><strong>Language:</strong> ${pageData.language}</p>
		<h3>Key Features:</h3>
		<ul>
			${pageData.features.map(f => `<li>${f}</li>`).join('')}
		</ul>
	</body>
	</html>
	`;
	
	res.end(html);
});

const PORT = 3001;
server.listen(PORT, () => {
	console.log(`Server starting on http://localhost:${PORT}...`);
});

File: basics/go-programs/09_html_rendering.go
package main

import (
	"fmt"
	"html/template"
	"net/http"
)

/*
	GO PROGRAM 9: HTML RENDERING & TEMPLATES
	Concepts:
	- 'html/template' is Go's template engine for safe and dynamic HTML.
	- Passing data (structs) to templates to render views.
*/

// PageData struct for our template
type PageData struct {
	Title    string
	Author   string
	Language string
	Features []string
}

func indexHandler(w http.ResponseWriter, r *http.Request) {
	// 1. Create data to pass to the template
	data := PageData{
		Title:    "Go Programming Language",
		Author:   "Google",
		Language: "Go (Golang)",
		Features: []string{"Concurrency", "Garbage Collection", "Simple Syntax", "Fast Compilation"},
	}
	
	// 2. Define the HTML Template
	tmplStr := `
	<!DOCTYPE html>
	<html>
	<head>
		<title>{{.Title}}</title>
		<style>
			body { font-family: Arial, sans-serif; padding: 2em; background-color: #f4f4f4; }
			h1 { color: #007d9c; }
			ul { color: #333; }
		</style>
	</head>
	<body>
		<h1>{{.Title}}</h1>
		<p><strong>Developed by:</strong> {{.Author}}</p>
		<p><strong>Language:</strong> {{.Language}}</p>
		<h3>Key Features:</h3>
		<ul>
			{{range .Features}}
				<li>{{.}}</li>
			{{end}}
		</ul>
	</body>
	</html>
	`
	
	// 3. Parse and execute the template
	tmpl, err := template.New("index").Parse(tmplStr)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	
	// Result is written directly to the ResponseWriter
	tmpl.Execute(w, data)
}

func main() {
	http.HandleFunc("/", indexHandler)
	fmt.Println("Server starting on http://localhost:8081...")
	http.ListenAndServe(":8081", nil)
}
